components/director/internal/domain/formationassignment/service.go

```diff
@@ -1081,21 +1047,108 @@ type AssignmentErrorWrapper struct {
 	Error AssignmentError `json:"error"`
 }
 
-func newNotificationStatusReportFromWebhookResponse(response *webhookdir.Response) *statusreport.NotificationStatusReport {
+func newNotificationStatusReportFromWebhookResponse(response *webhookdir.Response, operation model.FormationOperation, webhookMode graphql.WebhookMode) *statusreport.NotificationStatusReport {
 	var configuration json.RawMessage
 	if response.Config != nil {
 		configuration = []byte(*response.Config)
 	}
 
-	var state string
-	if response.State != nil {
-		state = *response.State
-	}
-
 	var errorFromResponse string
 	if response.Error != nil && *response.Error != "" {
 		errorFromResponse = *response.Error
 	}
 
-	return statusreport.NewNotificationStatusReport(configuration, state, errorFromResponse)
+	return statusreport.NewNotificationStatusReport(configuration, calculateState(response, operation, webhookMode), errorFromResponse)
+}
+
+func calculateState(response *webhookdir.Response, operation model.FormationOperation, webhookMode graphql.WebhookMode) string {
+	if response.State != nil && *response.State != "" {
+		return *response.State
+	}
+
+	if response.Error != nil && *response.Error != "" {
+		if operation == model.AssignFormation {
+			return string(model.CreateErrorAssignmentState)
+		}
+
+		return string(model.DeleteErrorAssignmentState)
+	}
+
+	if *response.ActualStatusCode == *response.SuccessStatusCode && webhookMode != graphql.WebhookModeAsyncCallback {
+		return string(model.ReadyAssignmentState)
+	}
+	if operation == model.AssignFormation && webhookMode == graphql.WebhookModeAsyncCallback {
```

Comment:
Maybe add a comment that this state shouldn't reach `Update` or `UpdateWithConstraints`?

Reply:
A comment was added to explain why we are skipping further notification processing. Also the `calculateState` method cuold be used in many places and for some of them this may not be the case. Generally whether the state shoud or should not reach certain places in the code does not concern the method that caculates the state.